cmds/ip/ip.go

```diff
@@ -66,25 +69,86 @@ func dev() netlink.Link {
 	return iface
 }
 
-func showips() {
+func showLinks(withAddresses bool) {
 	ifaces, err := netlink.LinkList()
 	if err != nil {
 		log.Fatalf("Can't enumerate interfaces? %v", err)
 	}
+	var buffer bytes.Buffer
+
 	for _, v := range ifaces {
-		addrs, err := netlink.AddrList(v, netlink.FAMILY_ALL)
-		if err != nil {
-			log.Printf("Can't enumerate addresses")
+		l := v.Attrs()
+		buffer.WriteString(fmt.Sprintf("%d: %s: <%s> mtu %d state %s\n", l.Index, l.Name,
+			strings.Replace(strings.ToUpper(fmt.Sprintf("%s", l.Flags)), "|", ",", -1),
+			l.MTU, strings.ToUpper(l.OperState.String())))
+		buffer.WriteString(fmt.Sprintf("    link/%s %s\n", l.EncapType, l.HardwareAddr))
+		if withAddresses {
+			formatLinkAddresses(&buffer, v)
 		}
-		log.Printf("%v: %v", v, addrs)
 	}
+	fmt.Print(buffer.String())
+}
+
+func formatLinkAddresses(buffer *bytes.Buffer, link netlink.Link) {
+	addrs, err := netlink.AddrList(link, netlink.FAMILY_ALL)
+	if err != nil {
+		log.Printf("Can't enumerate addresses")
+	}
+
+	for _, addr := range addrs {
+
+		var scope string
+		switch addr.Scope {
+		case int(netlink.SCOPE_UNIVERSE):
+			scope = "global"
+		case int(netlink.SCOPE_HOST):
+			scope = "host"
+		case int(netlink.SCOPE_SITE):
+			scope = "site"
+		case int(netlink.SCOPE_LINK):
+			scope = "link"
+		}
+
+		var inet string
+		if getIPVersion(addr) == 4 {
+			inet = "inet"
+		} else {
+			inet = "inet6"
+		}
+
+		buffer.WriteString(fmt.Sprintf("    %s %s", inet, addr.Peer))
+		if addr.Broadcast != nil {
+			buffer.WriteString(fmt.Sprintf(" brd %s", addr.Broadcast))
+		}
+		buffer.WriteString(fmt.Sprintf(" scope %s %s\n", scope, addr.Label))
+		var validLft, preferredLft string
+		if addr.PreferedLft == math.MaxUint32 {
+			preferredLft = "forever"
+		} else {
+			preferredLft = fmt.Sprintf("%dsec", addr.PreferedLft)
+		}
+		if addr.ValidLft == math.MaxUint32 {
+			validLft = "forever"
+		} else {
+			validLft = fmt.Sprintf("%dsec", addr.ValidLft)
+		}
+		buffer.WriteString(fmt.Sprintf("       valid_lft %s preferred_lft %s\n", validLft, preferredLft))
+	}
+}
+
+func getIPVersion(addr netlink.Addr) int {
+	// Okay... We can probably do something better
```

Comment:
:-)
This will do for now but is there info in the netlink.Addr?

Reply:
I didn't find anything inside `netlink.Addr` to get this information out of the box. This is why I did this poor implementation

Reply:
I believe len(addr.IPNet.IP) == 4 or 16 should tell you whether it's IPv4 or IPv6 (or, if you are looking at the peer, len(addr.Peer.IP))

Reply:
If you want to include IPv4 addresses represented in IPv6, you can check addr.IPNet.IP.To4(). To4() == nil means it's definitely IPv6. See https://github.com/golang/go/blob/master/src/net/ip.go#L186